airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/integration/test_report_based_streams.py

```diff
@@ -0,0 +1,423 @@
+        http_mocker.get(
+            _download_document_request(_DOCUMENT_DOWNLOAD_URL).build(),
+            _download_document_response(stream_name, data_format=data_format),
+        )
+
+        output = self._read(stream_name, config())
+        assert len(output.records) == DEFAULT_EXPECTED_NUMBER_OF_RECORDS
+
+    @pytest.mark.parametrize(("stream_name", "data_format"), STREAMS)
+    @HttpMocker()
+    def test_given_compressed_report_when_read_then_return_records(
+        self, stream_name: str, data_format: str, http_mocker: HttpMocker
+    ) -> None:
+        mock_auth(http_mocker)
+
+        http_mocker.post(_create_report_request(stream_name).build(), _create_report_response())
+        http_mocker.get(_check_report_status_request(_REPORT_ID).build(), _check_report_status_response(stream_name))
+        http_mocker.get(
+            _get_document_download_url_request(_REPORT_DOCUMENT_ID).build(), _get_document_download_url_response(compressed=True)
+        )
+
+        # a workaround to pass compressed document to the mocked response
+        document_request = _download_document_request(_DOCUMENT_DOWNLOAD_URL).build()
+        document_response = _download_document_response(stream_name, data_format=data_format, compressed=True)
+        document_request_matcher = HttpRequestMatcher(document_request, minimum_number_of_expected_match=1)
+        http_mocker._matchers.append(document_request_matcher)
+
+        http_mocker._mocker.get(
+            requests_mock.ANY,
+            additional_matcher=http_mocker._matches_wrapper(document_request_matcher),
+            response_list=[{"content": document_response.body, "status_code": document_response.status_code}],
+        )
+
+        output = self._read(stream_name, config())
+        assert len(output.records) == DEFAULT_EXPECTED_NUMBER_OF_RECORDS
+
+    @pytest.mark.parametrize(("stream_name", "data_format"), STREAMS)
+    @HttpMocker()
+    def test_given_http_status_500_then_200_when_create_report_then_retry_and_return_records(
+        self, stream_name: str, data_format: str, http_mocker: HttpMocker
+    ) -> None:
+        mock_auth(http_mocker)
+        http_mocker.post(
+            _create_report_request(stream_name).build(),
+            [response_with_status(status_code=HTTPStatus.INTERNAL_SERVER_ERROR), _create_report_response()],
+        )
+        http_mocker.get(_check_report_status_request(_REPORT_ID).build(), _check_report_status_response(stream_name))
+        http_mocker.get(_get_document_download_url_request(_REPORT_DOCUMENT_ID).build(), _get_document_download_url_response())
+        http_mocker.get(
+            _download_document_request(_DOCUMENT_DOWNLOAD_URL).build(),
+            _download_document_response(stream_name, data_format=data_format),
+        )
+
+        output = self._read(stream_name, config())
+        assert len(output.records) == DEFAULT_EXPECTED_NUMBER_OF_RECORDS
```

Comment:
I added this assertion so the reader of this test knows that after retry the records have been read even if the reader doesn't know much about how the framework works. If you think it's redundant, please let me know and I'll delete it.

Reply:
If we change the logic of how these streams work, it will either impact the rest of the tests (if it's a change of core functionality, like retrieving reports) and we will have to align our tests or these tests won't be impacted at all, and new tests will need to be added (for the new functionality). I'd like to leave this assertion here just to emphasise the records are still being read after an error received on the first attempt.